Clarify semantics of UTxO consistency check in `balanceTx`.

jonathanknowles opened the pull request:
## Issue

ADP-3185

## Description

This PR clarifies the semantics of the `balanceTx` function's UTxO consistency check.

In particular, it:
- redefines `guardWalletUTxOConsistencyWith` in terms of `Map.conflictsWith` — a function with documented semantics and property test coverage.
- revises the comment for `guardWalletUTxOConsistencyWith` to remove outdated references to `cardano-api`.

This allows us to:
- simplify the definition of `UTxOIndex`, removing the now unused `walletUTxO` field.

jonathanknowles committed:
Add functions `Data.Map.Strict.{conflicts,conflictsWith}`.

Anviking commented on lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs:
```diff
@@ -876,32 +875,18 @@ balanceTransactionWithSelectionStrategyAndNoZeroAdaAdjustment
+    -- | Ensure the wallet UTxO is consistent with the given UTxO.
     --
     -- They are not consistent iff an input can be looked up in both UTxO sets
     -- with different @Address@, or @TokenBundle@ values.
```
> iff an input can be looked up in both UTxO sets with different Address, or TokenBundle values.

It looks like this is no longer respected. 

The point of it is to allowing calling the `transactions-balance` API endpoint with inputs in the wallet which have been extended with datums or inlines scripts, even though the wallet is unable tracking them correctly.

Or said differently, this ensures procedures like
```haskell
setupAndUseReferenceScriptAtAddress :: Address -> IO ()
setupAndUseReferenceScriptAtAddress addr = do
    1. Have wallet create an inline script at `addr`
    2. Reference the input and use the script, calling transactions-balance manually providing the reference utxo for lookup
```

works for all values of `addr` (as opposed to breaking if `addr` belongs to the wallet.)

jonathanknowles replied:
@Anviking Okay -- I've adjusted the code so that two outputs `o1` and `o2` are deemed to be in conflict if (and only if) `toWalletTxOut o1 /= toWalletTxOut o2` (i.e., their `Address` or `TokenBundle` parts are different):

```hs
conflicts :: UTxO era -> UTxO era -> Map TxIn (TxOut era, TxOut era)
conflicts = Map.conflictsWith ((/=) `on` toWalletTxOut era) `on` unUTxO
```

jonathanknowles committed:
Use `Map.conflictsWith` to simplify `guardWalletUTxOConsistencyWith`.

jonathanknowles committed:
Remove outdated parts of comment for `guardWalletUTxOConsistencyWith`.

jonathanknowles committed:
Remove unneeded pattern match in `guardWalletUTxOConsistencyWith`.

jonathanknowles committed:
Remove unused `walletUTxO` field from `UTxOIndex` type.

This field is no longer used, so we can remove it.

jonathanknowles committed:
Inline definition of `walletUTxO` within `constructUTxOIndex`.

Since there's no longer a `walletUTxO` field, we no longer need a
separate definition.

Anviking reviewed (approved):
LGTM, thanks!
